core/unix/module_elf.c

```diff
@@ -1553,3 +1575,189 @@ module_relocate_rela(app_pc modbase, os_privmod_data_t *pd, ELF_RELA_TYPE *start
+        offs = ALIGN_BACKWARD(sec_hdr[elf_hdr->e_shstrndx].sh_offset, PAGE_SIZE);
+        str_size = ALIGN_FORWARD(sec_hdr[elf_hdr->e_shstrndx].sh_offset +
+                                     sec_hdr[elf_hdr->e_shstrndx].sh_size,
+                                 PAGE_SIZE) -
+            offs;
+        str_map =
+            os_map_file(fd, &str_size, offs, NULL, MEMPROT_READ, MAP_FILE_COPY_ON_WRITE);
+        if (str_map == NULL)
+            goto module_init_rseq_cleanup;
+        strtab = (char *)(str_map + sec_hdr[elf_hdr->e_shstrndx].sh_offset - offs);
+    }
+    bool found_array = false;
+    uint i;
+    ELF_SECTION_HEADER_TYPE *sec_hdr_start = sec_hdr;
+    /* The section entries on disk need load_offs.  The rseq entries in memory are
+     * relocated and only need the offset if relocations have not yet been applied.
+     */
+    ssize_t entry_offs = 0;
+    if (at_map || (DYNAMO_OPTION(early_inject) && !dr_api_entry && !dynamo_started))
+        entry_offs = load_offs;
+    for (i = 0; i < elf_hdr->e_shnum; i++) {
+#    define RSEQ_PTR_ARRAY_SEC_NAME "__rseq_cs_ptr_array"
+        if (strcmp(strtab + sec_hdr->sh_name, RSEQ_PTR_ARRAY_SEC_NAME) == 0) {
+            found_array = true;
+            byte **ptrs = (byte **)(sec_hdr->sh_addr + load_offs);
+            int j;
+            for (j = 0; j < sec_hdr->sh_size / sizeof(ptrs); ++j) {
+                /* We require that the table is loaded.  If not, bail. */
+                if (ptrs > (byte **)ma->end)
+                    goto module_init_rseq_cleanup;
+                /* We assume this is a full mapping and it's safe to read the data
+                 * (a partial map shouldn't make it to module list processing).
+                 */
+                rseq_process_entry((struct rseq_cs *)(*ptrs + load_offs), entry_offs);
+                ++ptrs;
+            }
+            break;
+        }
+        ++sec_hdr;
+    }
+    if (!found_array) {
+        sec_hdr = sec_hdr_start;
+        for (i = 0; i < elf_hdr->e_shnum; i++) {
+#    define RSEQ_SEC_NAME "__rseq_cs"
+#    define RSEQ_OLD_SEC_NAME "__rseq_table"
+            if (strcmp(strtab + sec_hdr->sh_name, RSEQ_SEC_NAME) == 0 ||
+                strcmp(strtab + sec_hdr->sh_name, RSEQ_OLD_SEC_NAME) == 0) {
+                /* There may be padding at the start of the section, so ensure we skip
+                 * over it.  We're reading the loaded data, not the file, so it will
+                 * always be aligned.
+                 */
+#    define RSEQ_CS_ALIGNMENT 4 * sizeof(__u64)
+                struct rseq_cs *array = (struct rseq_cs *)ALIGN_FORWARD(
+                    sec_hdr->sh_addr + load_offs, RSEQ_CS_ALIGNMENT);
+                int j;
+                for (j = 0; j < sec_hdr->sh_size / sizeof(*array); ++j) {
+                    /* We require that the table is loaded.  If not, bail. */
+                    if (array > (struct rseq_cs *)ma->end)
+                        goto module_init_rseq_cleanup;
```

Comment:
Do you think it's worthwhile adding some log messages when there's a goto-cleanup statement? It might be hard to debug an rseq initialization issue without manually adding such logs, and it seems prudent to me to just add them in the first place. I don't feel strongly, lots of the codebase does this so maybe it's the prevailing pattern.

Reply:
Yes, agreed.  Added in c12595fcadde2e2f6b6f06432d6015a37fbd1d04 along with turning this kind of error into a release-build-visible fatal error.